File: programmers_java/src/다단계칫솔판매.java
import java.util.HashMap;

public class 다단계칫솔판매 {
    class Solution {
        public int[] solution(String[] enroll, String[] referral, String[] seller, int[] amount) {
            int n = enroll.length;
            HashMap<String,Integer> nameToNum = new HashMap<String,Integer>();
            int[] sold = new int[n];
            int[] parent = new int[n];
            for(int i=0;i<n;i++){nameToNum.put(enroll[i],i);}
            for(int i=0;i<n;i++){
                if(!referral[i].equals("-"))parent[i] = nameToNum.get(referral[i]);
                else parent[i] = -1;
            }
            for(int i=0;i<seller.length;i++){giveFeedback(nameToNum.get(seller[i]),amount[i],sold,parent);}
            return sold;
        }
        void giveFeedback(int me,int amountLeft,int[] sold,int[] parent){
            int toParent = amountLeft/10;
            int mine = amountLeft - toParent;
            sold[me] += mine;
            if(toParent>0 && parent[me]!=-1) giveFeedback(parent[me],toParent,sold,parent);
        }
    }
}

File: programmers/src/main/kotlin/다단계칫솔판매.kt
fun main(){
    class Solution {
        fun solution(enroll: Array<String>, referral: Array<String>, seller: Array<String>, amount: IntArray): IntArray {
            val n = enroll.size
            val nameToNum = hashMapOf<String,Int>().apply { enroll.forEachIndexed{num,name-> this[name] = num} }
            val sold = IntArray(n)
            val parent = IntArray(n){-1}
            referral.forEachIndexed { myIndex,parentName -> if(parentName!="-") { parent[myIndex] = nameToNum[parentName]!! } }
            fun giveFeedback(me:Int,amountLeft:Int){
                val toParent = amountLeft/10
                val mine = amountLeft - toParent
                sold[me] += mine
                if(toParent>0 && parent[me]!=-1) giveFeedback(parent[me],toParent)
            }
            for(i in seller.indices){ giveFeedback(nameToNum[seller[i]]!!,amount[i]*100) }
            return sold
        }
    }
}